query/src/main/scala/filodb/query/util/HierarchicalQueryExperience.scala

```diff
@@ -0,0 +1,192 @@
+          filterTags,
+          datasetOptions.metricColumn)
+        if (isInclude) {
+          isHigherLevelAggregationApplicableWithIncludeTags(updatedShardKeyColumns, filterTags, tags)
+        }
+        else {
+          isHigherLevelAggregationApplicableWithExcludeTags(updatedShardKeyColumns, filterTags, tags)
+        }
+    }
+  }
+
+  /**
+   * @param metricColumnFilter
+   * @param metricSuffix
+   * @param filters
+   * @return
+   */
+  def getNextLevelAggregatedMetricName(metricColumnFilter: String, metricSuffix: String,
+                                       filters: Seq[ColumnFilter]): Option[String] = {
+    val metricNameSeq = LogicalPlan.getColumnValues(filters, metricColumnFilter)
+    metricNameSeq match {
+      case Seq() => None
+      // TODO: make it a config
+      case _ => Some(metricNameSeq.head.replaceFirst(":::.*", ":::" + metricSuffix))
+    }
+  }
+
+  /**
+   * Checks if the given aggregation operator is allowed for next level aggregated metric.
+   * @param operatorName - String - Aggregation operator name. Ex - sum, avg, min, max, count.
+   * @return - Boolean
+   */
+  def isAggregationOperatorAllowed(operatorName: String): Boolean = GlobalConfig.hierarchicalConfig match {
+    case Some(hierarchicalConfig) => hierarchicalConfig
+      .getStringList("allowed-aggregation-operators")
+      .contains(operatorName)
+    case None => false
+  }
+
+  /**
+   * Checks if the query with the given range function is allowed for next level aggregated metric.
+   * @param rangeFunctionEntryName - String - Range function name. Ex - rate, increase etc.
+   * @return - Boolean
+   */
+  def isRangeFunctionAllowed(rangeFunctionEntryName: String): Boolean = GlobalConfig.hierarchicalConfig match {
+    case Some(hierarchicalConfig) =>
+      hierarchicalConfig
+        .getStringList("allowed-range-functions")
+        .contains(rangeFunctionEntryName)
+    case None => false
+  }
+
+  /**
+   * Checks if at least one of the parent LogicalPlan(s) is allowed to update/optimize the metric name. If there are
+   * No parent LogicalPlans, then it returns false
+   * @param parentLogicalPlans - Seq[String] - List of parent LogicalPlans. Ex - Seq("BinaryJoin", "Aggregate")
+   * @return - Boolean
+   */
+  def isParentPeriodicSeriesPlanAllowedForRawSeriesUpdateForHigherLevelAggregatedMetric(
+                                       parentLogicalPlans: Seq[String]): Boolean = {
+    GlobalConfig.hierarchicalConfig match {
+      case Some(hierarchicalConfig) =>
+        parentLogicalPlans.exists(hierarchicalConfig
+          .getStringList("allowed-parent-logical-plans").contains)
+      case None => false
+    }
+  }
+
+  /**
+   * Checks if the PeriodicSeriesPlan which has access to RawSeries is allowed to update/optimize the metric name
+   * @param logicalPlanName - PeriodicSeriesPlan name. Ex - PeriodicSeriesWithWindowing, PeriodicSeries
```

Comment:
I think this would be tricky since we don't want to make the raw series update decision based on if it is a `PeriodicSeriesWithWindowing` or `PeriodicSeries` ( both of them are leaf level periodic series plan with access to RawSeries )

Reply:
Keeping the configs separate for now.

Reply:
Let me know if you want to sync on this offline and we can brainstorm together